modal/_serialization.py

```diff
@@ -42,13 +46,142 @@ def persistent_load(self, pid):
+                more_trailers=obj.get("more_trailers"),
+            )
+        )
+    elif msg_type == "http.disconnect":
+        return api_pb2.Asgi(http_disconnect=api_pb2.Asgi.HttpDisconnect())
+    else:
+        logger.debug("skipping serialization of unknown ASGI message type %r", msg_type)
+        return api_pb2.Asgi()
+
+
+def _deserialize_asgi(asgi: api_pb2.Asgi) -> Any:
+    def unflatten_headers(obj):
+        return list(zip(obj[::2], obj[1::2]))
+
+    msg_type = asgi.WhichOneof("type")
+    if msg_type == "http":
+        return {
+            "type": "http",
+            "http_version": asgi.http.http_version,
+            "method": asgi.http.method,
+            "scheme": asgi.http.scheme,
+            "path": asgi.http.path,
+            "query_string": asgi.http.query_string,
+            "headers": unflatten_headers(asgi.http.headers),
+            "client": (asgi.http.client_host, asgi.http.client_port) if asgi.http.HasField("client_host") else None,
+            "extensions": {
+                "http.response.trailers": {},
+            },
+        }
+    elif msg_type == "http_request":
+        return {
+            "type": "http.request",
+            "body": asgi.http_request.body,
+            "more_body": asgi.http_request.more_body,
+        }
+    elif msg_type == "http_response_start":
+        return {
+            "type": "http.response.start",
+            "status": asgi.http_response_start.status,
+            "headers": unflatten_headers(asgi.http_response_start.headers),
+            "trailers": asgi.http_response_start.trailers,
+        }
+    elif msg_type == "http_response_body":
+        return {
+            "type": "http.response.body",
+            "body": asgi.http_response_body.body,
+            "more_body": asgi.http_response_body.more_body,
+        }
+    elif msg_type == "http_response_trailers":
+        return {
+            "type": "http.response.trailers",
+            "trailers": unflatten_headers(asgi.http_response_trailers.headers),
+            "more_trailers": asgi.http_response_trailers.more_trailers,
+        }
+    elif msg_type == "http_disconnect":
+        return {"type": "http.disconnect"}
+    else:
+        assert msg_type is None
+        return None
+
+
+def serialize_data_format(obj: Any, data_format: int) -> bytes:
+    """Similar to serialize(), but supports other data formats."""
+    if data_format == api_pb2.DATA_FORMAT_PICKLE:
+        return serialize(obj)
+    elif data_format == api_pb2.DATA_FORMAT_ASGI:
+        return _serialize_asgi(obj).SerializeToString()
+    else:
+        raise InvalidError(f"Unknown data format {data_format!r}")
+
+
+def deserialize_data_format(s: bytes, data_format: int, client) -> Any:
+    if data_format == api_pb2.DATA_FORMAT_UNSPECIFIED:
+        # TODO: Remove this after Modal client version 0.52, when the data_format field is always set.
+        return deserialize(s, client)
```

Comment:
We need to keep this for compatibility with clients before today, since they won't be aware to set data_format. I made sure to set `data_format` in all locations and tested the code.